### ISSUE TITLE
[RFC] Revisiting Meta Device Initialization with `reset_parameters()`

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#104187]
**Overview**
Implementing a correct meta device initialization is challenging with either leaking FSDP into the `nn.Module` definition or leaking FSDP's private functions/attributes (`_is_fsdp_flattened()`). See this [issue](https://github.com/pytorch/pytorch/issues/90465) for the full context.

One "supported" method for meta device initialization is to have the model's modules implement `reset_parameters()` or for the user to pass in a `param_init_fn: Callable[[nn.Module], None]`. However, the current implementation is problematic for two reasons:
- `nn.Module.to_empty()` is recursive. It will reset already-initialized parameters in the case of nested FSDP wrapping.
- Only calling `reset_parameters()` with the FSDP-wrapped module (and not submodules) means that FSDP expects `reset_parameters()` to reset submodules' parameters, which violates the method's expected contract and requires unnecessary code duplication (to recurse into submodules) in possibly every `reset_parameters()` implementation.
https://github.com/pytorch/pytorch/blob/b2277075b0fe5cf085d369a313863f64c6fb50c3/torch/distributed/fsdp/_init_utils.py#L815-L826

**Proposed Solution**
The proposed solution is to change FSDP's contract: **`nn.Module.reset_parameters()` or the user's custom `param_init_fn: Callable[[nn.Module], None]` should only initialize the given module's parameters and _not_ those of its submodules.** This means that for the `param_init_fn`, I likely expect it to condition on the `nn.Module` type.

FSDP will make sure to only call `reset_parameters()` or the custom `param_init_fn` once per module throughout construction time and only on modules with parameters (i.e. `len(list(module.parameters(recurse=False))) > 0`).

In the short term, we can emulate an `nn.Module.to_empty(recurse=False)` with something like the following to unblock `reset_parameters()`:
```
with torch.no_grad():
    for module in modules_to_materialize:
        # TODO: Emulate `module.to_empty(recurse=False)` until that is
        # explicitly supported
        named_params = [
            (param_name, param)
            for param_name, param in module.named_parameters(recurse=False)
        ]
        if len(named_params) == 0:
            continue
        for param_name, param in named_params:
            materialized_param = nn.Parameter(
                torch.empty_like(param, device=materialization_device)
            )
            delattr(module, param_name)
            setattr(module, param_name, materialized_param)
        module.reset_parameters()  # type: ignore[operator]
```
However, this does not resolve the issue for a custom `param_init_fn` since, to my understanding, `to_empty()` is the only public way to materialize a meta-device module. Without `to_empty(recurse=False)`, we would expect the user to do something similar to the code snippet above in the `param_init_fn` implementation. This gives some additional urgency to supporting `to_empty(recurse=False)`. For a custom `param_init_fn`, I am thinking of something like:
```
def param_init_fn(module: nn.Module):
    module = module.to_empty(device=materialization_device, recurse=False)
    if isinstance(module, ...):
        ...
    elif isinstance(module, ...):
        ....
    ...
```
The motivation for using a `param_init_fn` like this is if the FSDP user does not own the model code and cannot add a `reset_parameters()` method.

---

cc: @albanD @mikaylagawarecki on `to_empty(recurse=False)` (which mainly just needs to make the recursion to begin `_apply` optional)
https://github.com/pytorch/pytorch/blob/b2277075b0fe5cf085d369a313863f64c6fb50c3/torch/nn/modules/module.py#L799-L801

cc: @lessw2020 @HamidShojanazeri @abhi-mosaic @speediedan @conceptofmind @supriyogit @eric-mitchell on if this breaks any existing materialization code (feel free to tag whomever to which this may be relevant)

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @zhaojuanmao @mrshenli @rohan-varma

**Comment by awgu** [comment]
Hi @albanD. Should `reset_parameters()` be responsible for both parameters _and buffers_? I.e., should we check `len(list(module.parameters(recurse=False))) == 0 and len(list(module.buffers(recurse=False))) == 0` to `continue`?

**Comment by albanD** [comment]
Thanks for sending this proposal Andrew!

> The proposed solution is to change FSDP's contract: nn.Module.reset_parameters() or the user's custom param_init_fn: Callable[[nn.Module], None] should only initialize the given module's parameters and not those of its submodules.

Note that this is already the current behavior of the `reset_parameters()` methods in core. https://github.com/pytorch/pytorch/blob/b2277075b0fe5cf085d369a313863f64c6fb50c3/torch/nn/modules/linear.py#L103

> Should reset_parameters() be responsible for both parameters and buffers?

I think we should do it yes. This is what batchnorm is doing in core btw.

> param_init_fn

Could you clarify what this function is and how you expect it to be used?

> to_empty(recurse=False)

Yes we can just do that. Hopefully we will be able to change the necessary internal functions (like _apply) to make this happen without too much pain.


More generally, a few things that we will want to happen here as well:
- Update tutorials and docs to recommend implementing reset_parameters().
- Add reset_parameters() as a method on Module to be able to reserve the name and easily document what the behavior should be.
- Add testing for reset_parameters() method to exist.
- param_init_fn might be a good way to override default init? There have been quite a few discussions on how to do that. We might want to have a deeper conversation on this one if we want to make it "the blessed way".

**Comment by awgu** [comment]
> > param_init_fn

> Could you clarify what this function is and how you expect it to be used?

My understanding is that `param_init_fn` should be like a catch-all `reset_parameters()` that takes in any `module` and resets its parameters (like you said, a way to override the default init). This is to handle the case where the user does not own the model code and cannot add a `reset_parameters()` to the module when needed. cc: @rohan-varma 

In that sense, it should have the same contract as `reset_parameters()`. The only difference should be that instead of being a method on the `nn.Module`, it can be a stand-alone function that takes in an `nn.Module` and operates in-place.

The follow-ups on `reset_parameters()` sound good to me!

**Comment by albanD** [comment]
But what would prevent actually overriding the reset_parameters() method on that specific Module?

**Comment by awgu** [comment]
> But what would prevent actually overriding the reset_parameters() method on that specific Module?

Does this mean having the user add a `reset_parameters()` to that module in their code?
```
# User code:
from transformers import QuantumTransformer
def qtransformer_reset_parameters(...):
    ...

QuantumTransformer.reset_parameters = qtransformer_reset_parameters
model = QuantumTransformer(...)
fsdp_model = FSDP(model, ...)
```
(Otherwise, my concern was that the user may not own the source code file for the model, e.g. this `QuantumTransformer`.)

**Comment by albanD** [comment]
> Does this mean having the user add a reset_parameters() to that module in their code?

Yes!
Instead of doing
```
def my_custom_init(mod):
  # The param_init_fn
```
You can update the class as you mentioned above yes.
Like I think doing `torch.nn.Linear.reset_parameters = my_init` is an ok way to override default linear initialization.

**Comment by awgu** [comment]
https://github.com/pytorch/pytorch/pull/104189
